### ISSUE TITLE
Various rules dont work well in astro files

### ISSUE THREAD

**Issue by TheElegantCoding** [eslint-stylistic/eslint-stylistic#259]
### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

I opened issue #210 sorry for not describing the error I want to mention something about, @antfu eslint also in his package eslint-config [#146](https://github.com/antfu/eslint-config/issues/146), it is okay not supporting Astro, but I think this is more related to stylistic or maybe to eslint-astro

as [#146](https://github.com/antfu/eslint-config/issues/146), right now in Astro files are some mistakes

1. the rules `indent` and `jsx-indent`, provoke that when we have two HTML elements in an Astro file they indent like this

```html
<h1>Hello</h1>

    <header>Another</header>
```

The correct way should be this

```html
<h1>Hello</h1>

<header>Another</header>
```

2. the rule `style/jsx-one-expression-per-line` in the script tag makes a bug that shows this error I mentioned in the last #210 issue

![image](https://github.com/eslint-stylistic/eslint-stylistic/assets/70770072/c6e6115b-03c7-4fd5-883e-a1bd09cdc7b1)

I found someone who runs the same issues in @antfu eslint-config [#146](https://github.com/antfu/eslint-config/issues/146), he says for now to set up eslint-astro we need to turn off these rules, which it is not ideal

```js
      rules: {
        //This is necessary to force a correct indentation in Astro
        'style/indent': ['error', 2],
        'style/jsx-indent': 'off',
        'style/jsx-one-expression-per-line': 'off',
      }
```

i will put my astro template repo and also my eslint-config i made similar to @antfu config, let me kwon if more information is needed

look the astro.astro file in the astro template, for other reason the files inside src are no linter

### Reproduction

[eslint config](https://github.com/TheElegantCoding/eslint-config-elegant-coding), [astro template](https://github.com/TheElegantCoding/astro_template_webpage)

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests

**Comment by antfu** [comment]
Ideally, the jsx rules should not be aware of Astro; either they should not be enabled in Astro files, or there is something that could be improved in `astro-eslint-parser` or `eslint-plugin-astro`.

/cc @ota-meshi, what do you suggest to move forward?

**Comment by Daydreamer-riri** [comment]
Are there any new solutions here？

**Comment by fwextensions** [comment]
> Ideally, the jsx rules should not be aware of Astro

I don't know if they need to be aware of Astro, but the `jsx-indent` rule does seem to assume that the tags can never appear in column 0, which is where it's standard to put the outermost tags in Astro or HTML.  So in a simple file like this:

```html
---
const text = "hello, world";
---

<div>
  <p>{text}</p>
</div>
```

the rule returns this error (assuming tab indents):

```
Expected indentation of 1 tab character but found 0  @stylistic/jsx-indent
```

Is there any way to configure the allowed starting column for tags?

**Comment by itmayziii** [comment]
I'm personally of the opinion that including JSX in these default rules is a weird decision, especially if they are going to conflict with other non JSX frameworks.

That being said, if anyone encounters this issue, this is how I solved the problem.

In my `eslint.config.js`
```
{
  /*
   * The JSX rules provided by stylistic should not be enabled with Astro files.
   * https://github.com/eslint-stylistic/eslint-stylistic/issues/259
   */
  name: 'astro-no-jsx',
  files: ['**/*.astro'],
  rules: Object.entries(stylistic.configs['recommended-flat'].rules).reduce((acc, entry) => {
    const [key, value] = entry
    if (key.startsWith('@stylistic/jsx')) {
      acc[key] = 'off'
    } else {
      acc[key] = value
    }
    return acc
  }, {}),
}
```

**Comment by antfu** [comment]
I don't know enough about Astro, and I don't use it. I would count on the community to figure out a solution to either here or the parser side.

**Comment by SalahAdDin** [comment]
> I'm personally of the opinion that including JSX in these default rules is a weird decision, especially if they are going to conflict with other non JSX frameworks.
> 
> That being said, this is how I solved the problem.
> 
> In my `eslint.config.js`
> 
> [
>   // other configs
> 
>   stylistic.configs['recommended-flat'],
>   {
>     /*
>      * The JSX rules provided by stylistic should not be enabled with Astro files.
>      * https://github.com/eslint-stylistic/eslint-stylistic/issues/259
>      */
>     name: 'astro-no-jsx',
>     files: ['**/*.astro'],
>     rules: Object.entries(stylistic.configs['recommended-flat'].rules).reduce((acc, entry) => {
>       const [key, value] = entry
>       if (key.startsWith('@stylistic/jsx')) {
>         acc[key] = 'off'
>       } else {
>         acc[key] = value
>       }
>       return acc
>     }, {}),
>   }
> ]

They are not only `jsx` files:
```
G:\Projects\sites\central-mosque-bogota\frontend\src\components\Welcome.astro
   2:23  error  Strings must use singlequote                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
                            @stylistic/quotes
   2:44  error  Extra semicolon                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
 @stylistic/semi
   3:24  error  Strings must use singlequote                                                                                    
  @stylistic/quotes
  15:10  error  Strings must use singlequote                                                                                    
  @stylistic/quotes
  17:10  error  Strings must use singlequote                                                                                    
@stylistic/quotes
  24:12  error  Strings must use singlequote                                                                                    
@stylistic/quotes
  53:8   error  `  #background {    position: fixed;    top: 0;    left: 0;    width: 100%;    height: 100%;    z-index: -1;    filter: blur(100px);  }  #container {    font-family:      Inter, Roboto, "Helvetica Neue", "Arial Nova", "Nimbus Sans", Arial,      sans-serif;    height: 100%;  }  main {    height: 100%;    display: flex;    justify-content: center;  }  #hero {    display: flex;    align-items: start;    flex-direction: column;    justify-content: center;    padding: 16px;  }  h1 {    font-size: 22px;    margin-top: 0.25em;  }  #links {    display: flex;    gap: 16px;  }  #links a {    display: flex;    align-items: center;    padding: 10px 12px;    color: #111827;    text-decoration: none;    transition: color 0.2s;  }  #links a:hover {    color: rgb(78, 80, 86);  }  #links a svg {    height: 1em;    margin-left: 8px;  }  #links a.button {    color: white;    background: linear-gradient(83.21deg, #3245ff 0%, #bc52ee 100%);    box-shadow:      inset 0 0 0 1px rgba(255, 255, 255, 0.12),      inset 0 -2px 0 rgba(0, 0, 0, 0.24);    border-radius: 10px;  }  #links a.button:hover {    color: rgb(230, 230, 230);    box-shadow: none;  }  pre {    font-family:      ui-monospace, "Cascadia Code", "Source Code Pro", Menlo, Consolas,      "DejaVu Sans Mono", monospace;    font-weight: normal;    background: linear-gradient(14deg, #d83333 0%, #f041ff 100%);    -webkit-background-clip: text;    -webkit-text-fill-color: transparent;    background-clip: text;    margin: 0;  }  h2 {    margin: 0 0 1em;    font-weight: normal;    color: #111827;    font-size: 20px;  }  p {    color: #4b5563;    font-size: 16px;    line-height: 24px;    letter-spacing: -0.006em;    margin: 0;  }  code {    display: inline-block;    background:      linear-gradient(66.77deg, #f3cddd 0%, #f5cee7 100%) padding-box,      linear-gradient(155deg, #d83333 0%, #f041ff 18%, #f5cee7 45%) border-box;    border-radius: 8px;    border: 1px solid transparent;    padding: 6px 8px;  }  .box {    padding: 16px;    background: rgba(255, 255, 255, 1);    border-radius: 16px;    border: 1px solid white;  }  #news {    position: absolute;    bottom: 16px;    right: 16px;    max-width: 300px;    text-decoration: none;    transition: background 0.2s;    backdrop-filter: blur(50px);  }  #news:hover {    background: rgba(255, 255, 255, 0.55);  }  @media screen and (max-height: 368px) {    #news {      display: none;    }  }  @media screen and (max-width: 768px) {    #container {      display: flex;      flex-direction: column;    }    #hero {      display: block;      padding-top: 10%;    }    #links {      flex-wrap: wrap;    }    #links a.button {      padding: 14px 18px;    }    #news {      right: 16px;      left: 16px;      bottom: 2.5rem;      max-width: 100%;    }    h1 {      line-height: 1.5;    }  }` must be placed on a new line  @stylistic/jsx-one-expression-per-line

G:\Projects\sites\central-mosque-bogota\frontend\src\components\actions\ThemeSwitch.astro
   80:29  error  ` import {   applyTheme,   getNextTheme,   getStoredTheme,   resolveTheme,   type ThemeSetting, } from "@utils/theme.utils"; function updateToggle(setting: ThemeSetting): void {   const resolved = resolveTheme(setting);   const button = document.querySelector<HTMLButtonElement>("[data-theme-toggle]");   if (!button) return;   const icons = button.querySelectorAll<HTMLElement>("span.material-symbols-outlined, span.material-symbols-rounded, span.material-symbols-sharp");   const settingIndex = setting === "light" ? 0 : setting === "dark" ? 1 : 2;   icons.forEach((icon, index) => {     if (index === settingIndex) {       icon.classList.remove("hidden");     } else {       icon.classList.add("hidden");     }   });   let nextLabel = "Switch to light theme";   if (setting === "light") {     nextLabel = "Switch to dark theme";   } else if (setting === "dark") {     nextLabel = "Use system theme";   }   button.setAttribute("aria-label", nextLabel);   button.dataset.themeSetting = setting;   button.dataset.themeResolved = resolved; } function initThemeToggle(): void {   const initialSetting = getStoredTheme();   applyTheme(initialSetting);   updateToggle(initialSetting);   const mql = window.matchMedia("(prefers-color-scheme: dark)");   const handleSystemChange = (): void => {     const currentSetting = getStoredTheme();     if (currentSetting !== "system") return;     applyTheme("system");     updateToggle("system");   };   if (mql.addEventListener) {     mql.addEventListener("change", handleSystemChange);   } else {     // @ts-expect-error older browsers     mql.addListener(handleSystemChange);   }   const button = document.querySelector<HTMLButtonElement>("[data-theme-toggle]");   if (!button) return;   button.addEventListener("click", () => {     const current = getStoredTheme();     const next: ThemeSetting = getNextTheme(current);     applyTheme(next);     updateToggle(next);   }); } if (typeof window !== "undefined") {   if (document.readyState === "loading") {     document.addEventListener("DOMContentLoaded", () => {       initThemeToggle();     });   } else {     initThemeToggle();   } }` must be placed on a new line  @stylistic/jsx-one-expression-per-line
   82:1   error  More than 1 blank line not allowed                                                                             
@stylistic/no-multiple-empty-lines
  174:1   error  More than 1 blank line not allowed                                                                             
@stylistic/no-multiple-empty-lines
   81:1   error  Too many blank lines at the beginning of file. Max of 0 allowed                                                
@stylistic/no-multiple-empty-lines
  112:2   error  Expected indentation of 4 spaces but found 0                                                                   
@stylistic/indent
  121:2   error  Expected indentation of 2 spaces but found 0                                                                   
@stylistic/indent
  147:2   error  Expected indentation of 2 spaces but found 0                                                                   
@stylistic/indent
  169:2   error  Expected indentation of 2 spaces but found 0                                                                   
@stylistic/indent
  173:1   error  Too many blank lines at the end of file. Max of 0 allowed                                                      
@stylistic/no-multiple-empty-lines

```

It happens to:
```astro
---
import "@fontsource-variable/public-sans";
import "@fontsource-variable/plus-jakarta-sans";
import "@fontsource-variable/material-symbols-outlined";
import materialSymbolsOutlined from "@fontsource-variable/material-symbols-outlined/files/material-symbols-outlined-latin-wght-normal.woff2";

import "@styles/global.css";
---

<!-- ﷽ - In the name of Allah, the Most Gracious, the Most Merciful -->
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width" />
    <link rel="icon" type="image/svg+xml" href="/favicon.svg" />
    <link rel="icon" href="/favicon.ico" />
    <meta name="generator" content={Astro.generator} />
    <title>Astro Basics</title>

    <link
      rel="preload"
      as="font"
      type="font/woff2"
      href={materialSymbolsOutlined}
      crossorigin="anonymous"
    />
  </head>
  <body>
    <slot />
  </body>
</html>

<style>
  html,
  body {
    margin: 0;
    width: 100%;
    height: 100%;
  }
</style>
```

Stylistic does not detect content inside frontmatter properly.

**Comment by SalahAdDin** [comment]
So I disabled the following:
```ts
    rules: {
      "@stylistic/arrow-parens": "error",
      "@stylistic/jsx-one-expression-per-line": "warn",
      "@stylistic/member-delimiter-style": "warn",
      "@stylistic/multiline-ternary": "warn",
      "@stylistic/quotes": "warn",
      "@stylistic/semi": "warn",
    },
```

And the multiline ternary is the following, crazy cause I don't know how it should be formatted:
```ts
{
  isLink ? (
    <a
      href={disabled ? undefined : href}
      class={classes}
      aria-disabled={disabled || undefined}
      aria-label={ariaLabel}
      {...rest}
    >
      {renderContent()}
    </a>
  ) : (
    <button
      type={type}
      disabled={disabled}
      class={classes}
      aria-label={ariaLabel}
      {...rest}
    >
      {renderContent()}
    </button>
  )
}
```

**Comment by SalahAdDin** [comment]
We have to disable `"@stylistic/jsx-one-expression-per-line": "off",` since it creates empty lines wrapping _style_ and _script_ tags when fixing the issue:
```astro
---
import { cn } from '@utils/styles.utils'
import type { HTMLAttributes } from 'astro/types'

/**
 * @component Dialog
 * @title Dialog
 * @description Root wrapper that wires up `DialogTrigger`, `DialogContent`, and `DialogClose` behavior.
 *
 * @preview
 *
 * <Dialog>
 *   <DialogTrigger class="rounded-md border px-3 py-2">Open</DialogTrigger>
 *   <DialogContent>
 *     Dialog content
 *   </DialogContent>
 * </Dialog>
 *
 * @usage
 *
 * ```astro nocollapse
 * ---
 * import { Dialog, DialogContent, DialogTrigger } from "@components/actions/Dialog";
 * ---
 *
 * <Dialog>
 *   <DialogTrigger class="rounded-md border px-3 py-2">Open</DialogTrigger>
 *   <DialogContent>
 *     Dialog content
 *   </DialogContent>
 * </Dialog>
 * ```
 *
 * @examples
 *
 * ### Disable close interactions
 *
 * ```astro nocollapse
 * ---
 * import { Dialog, DialogContent, DialogTrigger } from "@components/actions/Dialog";
 * ---
 *
 * <Dialog closeOnBackdropClick={false} closeOnEsc={false} lockScroll={false}>
 *   <DialogTrigger class="rounded-md border px-3 py-2">Open</DialogTrigger>
 *   <DialogContent>
 *     Dialog content
 *   </DialogContent>
 * </Dialog>
 * ```
 */

interface Props extends HTMLAttributes<'div'> {
  closeOnBackdropClick?: boolean
  closeOnEsc?: boolean
  lockScroll?: boolean
}

const {
  class: className,
  closeOnBackdropClick = true,
  closeOnEsc = true,
  lockScroll = true,
  ...rest
} = Astro.props as Props;
---

<div
  class={cn('relative', className)}
  data-dialog-root
  data-close-on-backdrop-click={closeOnBackdropClick ? 'true' : 'false'}
  data-close-on-esc={closeOnEsc ? 'true' : 'false'}
  data-lock-scroll={lockScroll ? 'true' : 'false'}
  {...rest}
>
  <slot />
</div>

<script>










  import { initDialogs } from "./dialog.controller";

  const setup = () => initDialogs();

  document.addEventListener("DOMContentLoaded", setup);
  document.addEventListener("astro:after-swap", setup);










</script>

<style>










  .overflow-hidden {
    overflow: hidden;
  }










</style>

````

**Comment by 9romise** [comment]
I don't use Astro and I'm not familiar with it.
@SalahAdDin Rather than repeatedly sharing the rule you've turned off, a minimal reproduction would make it a lot easier for us to track down the root cause and possibly fix it.

**Comment by SalahAdDin** [comment]
> I don't use Astro, and I'm not familiar with it. [@SalahAdDin](https://github.com/SalahAdDin) Rather than repeatedly sharing the rule you've turned off, a minimal reproduction would make it a lot easier for us to track down the root cause and possibly fix it.

Not minimal, but my current project using Astro and ESLint: https://github.com/SalahAdDin/central-mosque-bogota/tree/main/frontend

Let me know whether you need a simpler one, and I will get some time to make it.

**Comment by smnbbrv** [comment]
Just tested things on `6.0.0-beta.4`. All the described issues just do not happen.

@9romise is that on purpose or we just get incredibly lucky? :)

**Comment by SalahAdDin** [comment]
> Just tested things on `6.0.0-beta.4`. All the described issues just do not happen.
> 
> [@9romise](https://github.com/9romise) is that on purpose or we just get incredibly lucky? :)

Well, it seems to fix the issues, and also, we have incompatibility issues with our editor format.

**Comment by TheElegantCoding** [comment]
> Just tested things on `6.0.0-beta.4`. All the described issues just do not happen.
> 
> [@9romise](https://github.com/9romise) is that on purpose or we just get incredibly lucky? :)

Can you share what configuration you use? Thank you

**Comment by SalahAdDin** [comment]
> > Just tested things on `6.0.0-beta.4`. All the described issues just do not happen.
> > [@9romise](https://github.com/9romise) is that on purpose or we just get incredibly lucky? :)
> 
> Can you share what configuration you use? Thank you

Checked and everything is working now, thank you!

**Comment by smnbbrv** [comment]
> Can you share what configuration you use? Thank you

@TheElegantCoding literally the very default one

**Comment by 9romise** [comment]
> Just tested things on `6.0.0-beta.4`. All the described issues just do not happen.
> 
> [@9romise](https://github.com/9romise) is that on purpose or we just get incredibly lucky? :)

Normally, the described issues shouldn’t exist, so I think it’s on purpose. :)

**Comment by SalahAdDin** [comment]
> > > Just tested things on `6.0.0-beta.4`. All the described issues just do not happen.
> > > [@9romise](https://github.com/9romise) is that on purpose or we just get incredibly lucky? :)
> > 
> > 
> > Can you share what configuration you use? Thank you
> 
> Checked and everything is working now, thank you!

Not yet, I found a new issue:
```sh
`  [data-theme-icon] {    @apply hidden;  }  [data-theme-toggle][data-theme-setting="light"] [data-theme-icon="light"] {    @apply block;  }  [data-theme-toggle][data-theme-setting="dark"] [data-theme-icon="dark"] {    @apply block;  }  [data-theme-toggle][data-theme-setting="system"] [data-theme-icon="system"] {    @apply block;  }` must be placed on a new lineeslint[@stylistic/jsx-one-expression-per-line](https://eslint.style/rules/jsx-one-expression-per-line)
```

In there:

<img width="1668" height="1159" alt="Image" src="https://github.com/user-attachments/assets/4887d6bb-8dca-475b-8e13-11edc044a2bb" />

**Comment by SalahAdDin** [comment]
Another issue here:
```sh
/home/LuisAlaguna/Projects/central-mosque-bogota/frontend/src/components/actions/Dialog/Dialog.astro
   80:1  error    More than 1 blank line not allowed                                                                                                                                                                                                     @stylistic/no-multiple-empty-lines
   96:1  error    More than 1 blank line not allowed                                                                                                                                                                                                     @stylistic/jsx-one-expression-per-line
  109:1  error    More than 1 blank line not allowed                                                                                                                                                                                                     @stylistic/no-multiple-empty-lines
  122:1  error    More than 1 blank line not allowed                                                                                                                                                                                                     @stylistic/no-multiple-empty-lines
```

But not visible in the editor:

<img width="1077" height="1570" alt="Image" src="https://github.com/user-attachments/assets/558f57ee-e12d-4f4a-a9d9-b77f53e5faeb" />

The issue here is that the linter is trying to fix the first error by adding multiple empty lines around the affected line, creating a new issue.

**Comment by SalahAdDin** [comment]
up!
